Allow additional control as to which tests are run

milljm opened the pull request:
Supplying a path to the tests spec file, now allows one to run
only the tests supplied in that tests file:

```
./run_tests --spec-file tests/kernels/simple_diffusion/tests
      OK kernels/simple_diffusion.test
-----------------------------------------------------
```

Or by providing a directory containing the tests file:

```
./run_tests --spec-file tests/kernels/simple_diffusion
      OK kernels/simple_diffusion.diffusion_200x200
      OK kernels/simple_diffusion.diffusion_100x100
      OK kernels/simple_diffusion.uniform_refine_4
      OK kernels/simple_diffusion.test
-----------------------------------------------------
```

You can further alter the search logic with -i (and --re) accordingly:

```
./run_tests --spec-file tests/kernels/simple_diffusion -i speedtests
      OK tests/kernels/simple_diffusion.diffusion_200x200
      OK tests/kernels/simple_diffusion.uniform_refine_4
      OK tests/kernels/simple_diffusion.diffusion_100x100
-----------------------------------------------------
```

Closes #11076

Commit:
Allow additional control as to which tests are run

Supplying a path to the tests spec file, now allows one to run
only the tests supplied in that tests file:

./run_tests --spec-file tests/kernels/simple_diffusion/tests
      OK kernels/simple_diffusion.test
-----------------------------------------------------

Or by providing a directory containing the tests file:

./run_tests --spec-file tests/kernels/simple_diffusion
      OK kernels/simple_diffusion.diffusion_200x200
      OK kernels/simple_diffusion.diffusion_100x100
      OK kernels/simple_diffusion.uniform_refine_4
      OK kernels/simple_diffusion.test
-----------------------------------------------------

You can further alter the search logic with -i (and --re) accordingly:

./run_tests --spec-file tests/kernels/simple_diffusion -i speedtests
      OK tests/kernels/simple_diffusion.diffusion_200x200
      OK tests/kernels/simple_diffusion.uniform_refine_4
      OK tests/kernels/simple_diffusion.diffusion_100x100
-----------------------------------------------------

Closes #11076

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/168310/) on 87e3dd2 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11077/site)

This comment will be updated on new commits.

milljm commented:
Forgot to add a unit tests...

milljm commented:
This does not work quite the way I want it to yet. Do not merge.

The issue is, that os.walk still recursively finds tests that I don't want it to find, when I supply an ABS path to the spec file.

Commit:
Limit os.walk recursion with --spec-file

os.walk was proceeding beyond the supplied spec file
location.

Added a unittest for --spec-file usage.

Closes #11076

Commit:
refactor spec-file options a bit

remove nested if statement.

Closes #11076

Commit:
Cause error when invalid --spec-file used

When supplying an incorrect path, exit with an error
instead of resulting to default operations.

Fix docstring description.

Closes #11076

brianmoose reviewed (approved):
works for me
